walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,99 @@
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // Cases
+    // Options for no attendees - should be available all day.
+    if (request.getAttendees().size() == 0) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+    // Duration longer than a day will return empty array.
+    if (request.getDuration() > TimeRange.getTimeInMinutes(23, 59)) {
+      return Arrays.asList();
+    }
+    // Check if there are availabilities for all attendees
+    List<String> attendees = new ArrayList<>();
+    attendees.addAll(request.getAttendees());
+    attendees.addAll(request.getOptionalAttendees());
+    Collection<TimeRange> allAttendeesAvailabilities = findAvailability(events, request, attendees);
+
+    // If there are no available timeslots for all attendees including optional attendees,
+    // Find availabilities for the mandatory attendees.
+    if (allAttendeesAvailabilities.size() == 0) {
+      return findAvailability(events, request, request.getAttendees());
+    }
+
+    return allAttendeesAvailabilities;
+  }
+
+  /** Private helper method to find all availabilities of people from the request. */
+  private Collection<TimeRange> findAvailability(
+      Collection<Event> events, MeetingRequest request, Collection<String> attendees) {
+    // Find all the unavailable times of the attendees.
+    List<TimeRange> unavailabilities = new ArrayList<>();
+    for (Event event : events) {
+      if (attendees.stream().anyMatch(s -> event.getAttendees().contains(s))) {
+        unavailabilities.add(event.getWhen());
+      }
+    }
+    // If there are no unavailable times, return whole day.
+    if (unavailabilities.size() == 0) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+    List<TimeRange> mergedUnavailabilities =
+        mergeAllUnavailabilities((ArrayList<TimeRange>) unavailabilities);
+
+    // Find all available times by checking if gaps between busy time ranges are greater than the
+    // duration.
+    List<TimeRange> availabilities = new ArrayList<>();
+    int start = TimeRange.START_OF_DAY;
+    long duration = request.getDuration();
+    for (TimeRange interval : mergedUnavailabilities) {
+      if (interval.start() - start >= duration) {
+        availabilities.add(TimeRange.fromStartEnd(start, interval.start(), false));
+      }
+      start = interval.end();
+    }
+    // Check if there is a lot between there and the end of the day.
+    if (TimeRange.END_OF_DAY - start >= duration) {
+      availabilities.add(TimeRange.fromStartEnd(start, TimeRange.END_OF_DAY, true));
+    }
+    return availabilities;
+  }
+
+  /**
+   * Private helper method using Stack to merge overlapping time ranges from an ArrayList.
+   * Merging code from https://www.geeksforgeeks.org/merging-intervals/
+   */
+  private List<TimeRange> mergeAllUnavailabilities(ArrayList<TimeRange> unavailabilities) {
+    Stack<TimeRange> unavailabilitiesStack = new Stack<>();
```

Comment:
The Stack class is quite old and is considered something to avoid. Google Java Practices suggests you use the ArrayDeque class instead, e.g.
```Deque<TimeRange> unavailabilitiesStack = new ArrayDeque<>();```

There are a couple of reasons for this:
- A Stack is nonstandard because it isn't part of the Java Collections Framework
- It is [slow because it is built to be inherently thread-safe](https://stackoverflow.com/questions/23908511/how-is-arraydeque-faster-than-stack)
- Some people think it has a strange API because it implements List;
- and a [few other reasons as well](https://stackoverflow.com/questions/12524826/why-should-i-use-deque-over-stack).

Interestingly, the `mergeAllUnavailabilities` function uses the fact that Stack implements the List interface, because it returns the Stack and just treats it as a List. This means that if you decide to use ArrayDeque, you'll have to convert it to a ArrayList or something before returning it, e.g.
```return new ArrayList<TimeRange>(unavailabilitiesStack);```

However, I think there is also a problem with this because the Stack class and the ArrayDeque class have opposite iteration orders. I'm guessing if you build an ArrayList like this, the order of the TimeRanges will be the opposite of before, and the order of the List is important in this algorithm. So you'll also probably have to reverse the order of the ArrayList, e.g.
```Collections.reverse(list);```

Quite a few changes to fix one thing 😂! I'm happy for you to leave this as a TODO and come back to it only if you have time. I saw the Stack class and thought it was a good point to raise, but then realised there was a bit more to it...

[Some notes on selecting Java collections if you are interested.](https://docs.google.com/document/d/1voHo_fTfX_DPcRVXrmkZyjjrgxeuuvcp2nfG4bBw2BA/edit)

Reply:
I've changed it to a Deque now! It definitely took a bit of time trying to figure out why some tests weren't passing and why I was getting errors but I eventually got there. Thanks for the heads up about the Stack class!